### ISSUE TITLE
SQLiteConnectionPool Issue with Latest Master

### ISSUE THREAD

**Issue by tikimcfee** [couchbase/couchbase-lite-android#697]
Hey there,

We're testing out the latest pull of master to try out some of the posted changes. Without changing any of our code (besides using the new transaction interface), we're frequently seeing this ...:
```
11-09 13:01:18.930 28724-28961/com.* W/SQLiteConnectionPool: The connection pool for database '/data/data/com.*/files/df923eb22b33.cblite2/db.sqlite3' has been unable to grant a connection to thread 745 (ChangeTracker-http://192.168.0.17:4984/sync_gateway_sw1) with flags 0x1 for 240.06902 seconds.
11-09 13:01:18.930 28724-28961/com.* W/SQLiteConnectionPool: Connections: 0 active, 1 idle, 0 available.
```
... which seems remain hung indefinitely when it occurs. On a rare occasion we can get replication to start and retrieve some documents, but it eventually fails to the above.

Is this a known issue, or, any way to move passed this to continue testing?

**Comment by hideki** [comment]
Hi @tikimcfee,
SQLite Connection pool issue might be caused by out-of-memory. I assume that replicated document size is huge or each document has many revisions. They could consume a lot of memory during replication. We might be able to reduce memory usage if you could help us.
Thanks!

**Comment by tikimcfee** [comment]
It's possible the replicator is hitting a number of documents in our test db that have a high revision count, and we've seen issues before with large revision counts before too. However, the memory allocator isn't posting more than a few MB of memory being used, so it seems really unlikely in this case.

**Comment by tikimcfee** [comment]
Whoops, forgot to ask: anything in mind you'd like us to try?

For more context, it seems like the batcher hits 100 or so docs before we see that message thrown.

**Comment by pasin** [comment]
http://tleyden.github.io/blog/2013/11/14/investigating-an-android-sqlite-threading-deadlock/
http://stackoverflow.com/questions/25563352/the-connection-pool-for-database-data-data-msv-database-db-has-been-unable-to

**Comment by pasin** [comment]
@tikimcfee does this happen when you pull a large document set? By the time it happens, can you describe what your application is doing (e.g. does it also do live query at the same time as it's pulling the document).

**Comment by tikimcfee** [comment]
@pasin Hey there. Our test replications targets regularly sync down 10k+ documents to device, but the lock up seems to most often occur before a single document comes down. Most cases show ~100 documents queued up (total changes / changes downloaded) in the download batcher at that point.

There is never a live query running, and only a single pull replicator is active while our application does an initial synchronization of core documents.

Our document use case is detailed in issues 686 and 687 pretty well if you want more details about them. TL:DR; small documents, lots of small 50kb attachments with a few large >1mb, potentially high revision counts on old documents, etc.

The path looks like this:
1. Create a pull replicator, set auth headers on it, point it at the server
2. Set as 'one shot' ('continuous' mode doesn't work currently as a method of determining when a full initial set of documents has come down as the idle/active states are inconsistent)
3. Start replication.

Replication usually starts up, hits a few queued documents, and then hangs on the above message.

**Comment by pasin** [comment]
@hideki could the issue be related to the fix that you recently worked on for the one-shot replication?

**Comment by hideki** [comment]
@pasin, I don't think so because I just submitted PR yesterday. But this ticket was submitted 23 days ago.

**Comment by pasin** [comment]
@hideki Sorry for confusion. I intended to ask whether the fix that you recent made could fix the issue related to this thread or not?

**Comment by hideki** [comment]
I think size of document, number of documents or number of revisions causes this issue. I hope avoiding Android SQLiteDatabase class and its JNI could solve this issue.

**Comment by pasin** [comment]
@tikimcfee would it be possible to share the test data so that I can try to reproduce the issue? Without being to reproduce the issue, it's hard to understand the root cause of the issue.

Also enabling some logging may help seeing what is going on underneath at that time. You could try to enable the following logging and share the log messages:

Manager.enableLogging(Log.TAG, Log.VERBOSE);
Manager.enableLogging(Log.TAG_SYNC, Log.VERBOSE);
Manager.enableLogging(Log.TAG_SYNC_ASYNC_TASK, Log.VERBOSE);
Manager.enableLogging(Log.TAG_BATCHER, Log.VERBOSE);
Manager.enableLogging(Log.TAG_CHANGE_TRACKER, Log.VERBOSE);
Manager.enableLogging(Log.TAG_DATABASE, Log.VERBOSE);

**Comment by pasin** [comment]
I think I have faced this issue myself today while working on #686. It happens while trying to pull large set of document and normally when reproducible, it happens right after start the pull replication. It seems to me that [1]  there is a race condition when begin and end nested transaction or [2] somehow the outer transaction doesn't end and hold the connection forever. 

I'm reading the code that handles begin and end transaction and I couldn't spot anything so I guessed the issue is the later case.

**Comment by pasin** [comment]
@zgramana I am marking this as a prep release issue because the issue could be reproduce easily and once it happens it completely hangs the puller replication.

**Comment by pasin** [comment]
The hang was caused by the following line never returns, as a result, the transaction doesn't get closed.
https://github.com/couchbase/couchbase-lite-java-core/blob/master/src/main/java/com/couchbase/lite/replicator/PullerInternal.java#L615

In pauseOrResume() method, there is a dead lock with calling batcher.count() as the batcher's inbox is locked while the batcher is processing but the batcher cannot finish processing as it can't obtain a database connection as the connection is held by the pending open transaction.

Another thing I have found that the batcher could process its inbox by 2 different threads: one by the thread that calls queueObjects when the inbox size > capacity and another by the work executor thread when delaying the processing. I believe by design the batcher should process its inbox only by using the work executor thread. I believe fixing this will also fix the dead lock above.
